source/slang-llvm/slang-llvm-filecheck.cpp

```diff
@@ -0,0 +1,189 @@
+    if(std::error_code err = rulesTextOrError.getError())
+    {
+        const std::string message = "Unable to load FileCheck rules file: " + err.message();
+        testReporter(userReporterData, TestMessageType::RunError, message.c_str());
+        return TestResult::Fail;
+    }
+    SmallString<4096> rulesBuffer;
+    StringRef rulesStringRef = fc.CanonicalizeFile(*rulesTextOrError.get(), rulesBuffer);
+    sourceManager.AddNewSourceBuffer(
+        MemoryBuffer::getMemBuffer(rulesStringRef, rulesFilePath),
+        SMLoc());
+
+    SmallString<4096> inputBuffer;
+    const auto inputStringMB = MemoryBuffer::getMemBuffer(
+        StringRef(stringToCheck->data, stringToCheck->count),
+        stringToCheckName,
+        false);
+    const StringRef inputStringRef = fc.CanonicalizeFile(*inputStringMB.get(), inputBuffer);
+    sourceManager.AddNewSourceBuffer(
+        MemoryBuffer::getMemBuffer(inputStringRef, stringToCheckName),
+        SMLoc());
+
+    // Initialize this with a 'RunError' failure type. We'll "downgrade" this to
+    // 'TestFailure' once we've done the FileCheck setup.
+    ReporterData reporterData{
+        testReporter,
+        userReporterData,
+        colorDiagnosticOutput,
+        programName,
+        TestMessageType::RunError};
+    sourceManager.setDiagHandler(fileCheckDiagHandler, static_cast<void*>(&reporterData));
+
+    auto checkPrefix = fc.buildCheckPrefixRegex();
+    if(fc.readCheckFile(sourceManager, rulesStringRef, checkPrefix))
+    {
+        // FileCheck failed to find or understand any FileCheck rules in
+        // the input file, automatic fail, and reported to the diag handler .
+        return TestResult::Fail;
+    }
+
+    // We've done the FileCheck setup, so make sure that any diagnostics
+    // reported on from here are just a regular test failure.
+    reporterData.testMessageType = TestMessageType::TestFailure;
+    if(!fc.checkInput(sourceManager, inputStringRef))
+    {
+        // An ordinary failure, the FileCheck rules didn't match
+        return TestResult::Fail;
+    }
+
+    return TestResult::Pass;
+}
+
+void* LLVMFileCheck::castAs(const Guid& guid)
+{
+    if (auto ptr = getInterface(guid))
+    {
+        return ptr;
+    }
+    return getObject(guid);
+}
+
+void* LLVMFileCheck::getInterface(const Guid& guid)
+{
+    if (guid == ISlangUnknown::getTypeGuid() ||
+        guid == ICastable::getTypeGuid() ||
+        guid == IFileCheck::getTypeGuid())
+    {
+        return static_cast<IFileCheck*>(this);
+    }
+    return nullptr;
+}
+
+void* LLVMFileCheck::getObject(const Guid& guid)
+{
+    SLANG_UNUSED(guid);
+    return nullptr;
+}
+
+} // namespace slang_llvm
+
+extern "C" SLANG_DLL_EXPORT SlangResult createLLVMFileCheck_V1(const SlangUUID& intfGuid, Slang::IFileCheck** out)
+{
+    Slang::ComPtr<slang_llvm::LLVMFileCheck> fileCheck(new slang_llvm::LLVMFileCheck);
+
+    if (auto ptr = fileCheck->castAs(intfGuid))
```

Comment:
Nitpick: The making the out parameter take a type, makes this seem a little odd. I can ask for an interface (via the guid), but I may get back something different and incompatible with the out type.

It might be better to make the out parameter void**, and then that would remove this confusion.

Reply:
I agree, I've corrected it and will make a separate PR to fix `createLLVMDownstreamCompiler`